### PR TITLE
CQL Long

### PR DESCRIPTION
**Pull request by cmoesel**
This PR introduces support for CQL Long, represented using `BigInt` in JavaScript. Since Integer and Decimal are represented using Number, we can easily distinguish CQL Longs from Integers and Decimals by their type (`typeof var === 'bigint'`). This actually made implementation easier (and more correct) compared to my initial implementation using Number (which I thought would be easier since it was more of an incremental change, but maybe not).

You can test this by building and running a CQL library that uses Long types -- or you can just review the unit tests and trust that they are properly exercising the capability.

**Submitter:**

- [x] This pull request describes why these changes were made
- [x] Code diff has been done and been reviewed (it does not contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] Code coverage has not gone down and all code touched or added is covered.
- [x] Code passes lint and prettier (hint: use `npm run test:plus` to run tests, lint, and prettier)
- [x] All dependent libraries are appropriately updated or have a corresponding PR related to this change (N/A)
- [x] `cql4browsers.js` built with `npm run build:browserify` if source changed.

**Reviewer:**

Name:

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: test/elm/literal/literal-test.ts [hunk 1]
```diff
@@ -31,6 +31,14 @@ describe('Literal', () => {
     (await this.intOne.exec(this.ctx)).should.equal(1);
   });
 
+  it('should convert 1L to 1', function () {
```
Comment (lmd59) [inline test/elm/literal/literal-test.ts]: Would this be a little more clear?
```suggestion
  it('should convert 1L to 1n', function () {
```

File: test/elm/literal/literal-test.ts [hunk 2]
```diff
@@ -31,6 +31,14 @@ describe('Literal', () => {
...
+    this.longOne.value.should.equal(1n);
+  });
+
+  it('should execute 1L as 1', async function () {
```
Comment (lmd59) [inline test/elm/literal/literal-test.ts]: Would this be a little more clear?
```suggestion
  it('should execute 1L as 1n', async function () {
```

File: test/elm/arithmetic/data.cql [hunk 3]
```diff
@@ -163,65 +199,143 @@ define SubtractUcum: (25 'km' - 5 'm') = 24995 'm'
...
+define LongSubtractNearUnderflow: minimum Long - 0L
+define LongMultiplyOverflow: maximum Long * 2L
+define LongMultiplyUnderflow: minimum Long * 2L
+define LongMultiplyNearOverflow: maximum Long * 1L
+define LongMultiplyNearUnderflow: minimum Long * 1L
+define LongDivideOverflow: maximum Long / 0.5
+define LongDivideUnderflow: minimum Long / 0.5
+define LongDivideNearOverflow: ToLong(maximum Integer) / 1L
+define LongDivideNearUnderflow: ToLong(minimum Integer) / 1L
+define LongDivideByZero: 1L / 0L
```
Comment (lmd59) [inline test/elm/arithmetic/data.cql:255]: Would it add anything to consider a long/integer mixed divide by zero?

File: test/elm/aggregate/data.cql [hunk 4]
```diff
@@ -121,6 +125,7 @@ define IncompatibleUnitsNull: PopulationStdDev({1 'mg/d', 0.002 '/d'})
 // @Test: Product
 define decimal_product: Product({1.0, 2.0, 3.0, 4.0})
 define integer_product: Product({5, 4, 5})
+define long_product: Product({5L, 4L, 5L})
```
Comment (lmd59) [inline test/elm/aggregate/data.cql:148]: Recommend adding a test for over/underflow on an aggregate product.

File: test/elm/aggregate/data.cql [hunk 5]
```diff
@@ -6,6 +6,7 @@ define is_null: Count(null as List<Integer>)
 
 // @Test: Sum
 define not_null: Sum({1,2,3,4,5})
+define longs: Sum({1L,2L,3L,4L,5L})
```
Comment (lmd59) [inline test/elm/aggregate/data.cql:13]: Good test for aggregate functionality. Would also recommend adding a test for over/underflow on an aggregate sum.

**Review by lmd59 — COMMENTED** [review]
Looking good! The tests are very thorough and helpful for confirming successful implementation. I left a couple of small comments on tests. Most are just nitpick/optional, but I think a couple more aggregate tests wouldn't hurt.

I also wanted to ask about support for Long in the interval datatype 
https://cql.hl7.org/04-logicalspecification.html#interval states that "An interval must be defined using a point type that supports comparison, as well as Successor and Predecessor operations, and Minimum and Maximum Value operations." Long seems like it would fit in this category. Is there any reason not to support Long as part of an interval?

File: test/elm/arithmetic/data.cql (hunk 3 above)
Comment (cmoesel) [reply test/elm/arithmetic/data.cql:255]: Honestly, I don't think so because the translator will just apply a conversion function so they both are the same type by the time the division happens.

**Comment by cmoesel** [comment]
Thanks for the excellent review! You uncovered some gaps in my implementation and also helped me to find more gaps on my own. But I will also note: _be careful what you wish for!_ I just added a lot more changes and tests.

* I added support for `Interval<Long>` in 79c0fb8, including support for multiple operations that work on intervals.
* I added tests for sum and product overflow/underflow and in the process discovered we weren't checking for overflow/underflow at all, so I also fixed the implementation. See: 6eaf011.
* This led me to realize that we could improve our type handling in general by using ELM result types when they are available, so I did that and fixed a bunch of related stuff in 4b5e62d. That one's kind of big.
* And finally, I found a couple more spots where we weren't explicitly accounting for long/bigint, so I fixed those in 9620bb8.

Thanks again for the great review. Sorry for adding a bunch more work for your review!

**Comment by cmoesel** [comment]
FYI: I rebased this branch on master to take in the latest `test-server` changes. There was minimal impact to the changes in this branch (just some small changes in `tsconfig.json`).

File: test/elm/arithmetic/arithmetic-test.ts [hunk 6]
```diff
@@ -829,13 +1211,30 @@ describe('OutOfBounds', () => {
...
+
     it('should return null for successor overflow', async function () {
       should(await this.decimalSuccessorOverflow.exec(this.ctx)).be.null();
     });
 
     it('should return null for predecessor underflow', async function () {
       should(await this.decimalPredecessorUnderflow.exec(this.ctx)).be.null();
     });
+
+    // NOTE: skipping successor/predecessor tests near overflow due to JS Number impreciision
```
Comment (lmd59) [inline test/elm/arithmetic/arithmetic-test.ts]:
```suggestion
    // NOTE: skipping successor/predecessor tests near overflow due to JS Number imprecision
```

File: test/elm/arithmetic/arithmetic-test.ts [hunk 7]
```diff
@@ -882,31 +1329,54 @@ describe('OutOfBounds', () => {
     it('should return null for predecessor underflow', async function () {
       should(await this.quantityPredecessorUnderflow.exec(this.ctx)).be.null();
     });
+
+    // NOTE: skipping successor/predecessor tests near overflow due to JS Number impreciision
```
Comment (lmd59) [inline test/elm/arithmetic/arithmetic-test.ts]:
```suggestion
    // NOTE: skipping successor/predecessor tests near overflow due to JS Number imprecision
```

File: test/elm/aggregate/data.cql [hunk 8]
```diff
@@ -119,11 +140,28 @@ define NumbersAndQuantities: PopulationStdDev({1 ,2 ,3 ,4 'ml',5 })
...
+define integers_above_max_value_product: Product({2, 1073741824}) // 2147483648
+define integers_at_min_value_product: Product({-1, 2, 1073741824}) // -2147483648
+define integers_below_min_value_product: Product({-2, 1073741825}) // -2147483650
+define long_product: Product({5L, 4L, 5L})
+define longs_at_max_value_product: Product({1L, 9223372036854775807L}) // 9223372036854775807
+define longs_above_max_value_product: Product({2L, 4611686018427387904L}) // 9223372036854775808
+define longs_at_min_value_product: Product({-1L, 2L, 4611686018427387904L}) // -9223372036854775808
+define longs_below_min_value_product: Product({-1L, 3L, 3074457345618258603L}) // -9223372036854775809
+define decimal_product: Product({1.0, 2.0, 3.0, 4.0})
+define decimals_at_max_value_product: Product({99999999999999999999.99999999})
```
Comment (lmd59) [inline test/elm/aggregate/data.cql]:
```suggestion
define decimals_at_max_value_product: Product({99999999999999999999.99999999, 1.0})
```
Optional: Not really necessary, but might make the test a little clearer

File: test/elm/aggregate/data.cql [hunk 9]
```diff
@@ -119,11 +140,28 @@ define NumbersAndQuantities: PopulationStdDev({1 ,2 ,3 ,4 'ml',5 })
...
+define decimals_above_max_value_product: Product({99999999999999999999.99999999, 2.0})
+define decimals_at_min_value_product: Product({-99999999999999999999.99999999})
```
Comment (lmd59) [inline test/elm/aggregate/data.cql]:
```suggestion
define decimals_at_min_value_product: Product({-99999999999999999999.99999999, 1.0})
```
Optional

File: test/elm/aggregate/data.cql [hunk 10]
```diff
@@ -119,11 +140,28 @@ define NumbersAndQuantities: PopulationStdDev({1 ,2 ,3 ,4 'ml',5 })
...
+define decimals_below_min_value_product: Product({-99999999999999999999.99999999, 2.0})
 define quantity_product: Product({1.0 'g', 2.0 'g', 3.0 'g', 4.0 'g'})
+define quantities_at_max_value_product: Product({99999999999999999999.99999999 'g'})
```
Comment (lmd59) [inline test/elm/aggregate/data.cql]:
```suggestion
define quantities_at_max_value_product: Product({99999999999999999999.99999999 'g', 1.0 'g'})
```
Optional

File: test/elm/aggregate/data.cql [hunk 11]
```diff
@@ -119,11 +140,28 @@ define NumbersAndQuantities: PopulationStdDev({1 ,2 ,3 ,4 'ml',5 })
...
+define quantities_above_max_value_product: Product({99999999999999999999.99999999 'g', 2.0 'g'})
+define quantities_at_min_value_product: Product({-99999999999999999999.99999999 'g'})
```
Comment (lmd59) [inline test/elm/aggregate/data.cql]:
```suggestion
define quantities_at_min_value_product: Product({-99999999999999999999.99999999 'g', 1.0 'g'})
```
Optional

File: test/elm/aggregate/aggregate-test.ts [hunk 12]
```diff
@@ -31,8 +31,93 @@ describe('Sum', () => {
     setup(this, data);
   });
 
-  it('should be able to sum lists without nulls', async function () {
-    (await this.not_null.exec(this.ctx)).should.equal(15);
+  it('should be able to sum lists with integers nulls', async function () {
```
Comment (lmd59) [inline test/elm/aggregate/aggregate-test.ts]:
```suggestion
  it('should be able to sum lists with integers', async function () {
```

File: src/datatypes/interval.ts [hunk 13]
```diff
@@ -497,6 +506,8 @@ export class Interval {
       let diff = Math.abs(highValue - lowValue);
       diff = Math.round(diff * Math.pow(10, 8)) / Math.pow(10, 8);
       return new Quantity(diff, closed.low.unit);
+    } else if (typeof closed.low === 'bigint') {
+      return closed.high >= closed.low ? closed.high - closed.low : closed.low - closed.high;
```
Comment (lmd59) [inline src/datatypes/interval.ts]: I'm not sure why we have the ternary here. Shouldn't it just be `closed.high - closed.low` according to https://cql.hl7.org/04-logicalspecification.html#width

File: src/datatypes/interval.ts [hunk 14]
```diff
@@ -528,6 +539,9 @@ export class Interval {
       let diff = Math.abs(highValue - lowValue) + pointSize.value;
       diff = Math.round(diff * Math.pow(10, 8)) / Math.pow(10, 8);
       return new Quantity(diff, closed.low.unit);
+    } else if (typeof closed.low === 'bigint') {
+      const diff = closed.high >= closed.low ? closed.high - closed.low : closed.low - closed.high;
```
Comment (lmd59) [inline src/datatypes/interval.ts]: Similar to above: I'm not sure why we have the ternary here. Shouldn't it just be closed.high - closed.low ()
according to https://cql.hl7.org/04-logicalspecification.html#size

File: src/datatypes/interval.ts [hunk 15]
```diff
@@ -528,6 +539,9 @@ export class Interval {
...
+      return diff + 1n;
```
Comment (lmd59) [inline src/datatypes/interval.ts]: Ideally 
```suggestion
      return diff + pointSize.value;
```
according to https://cql.hl7.org/04-logicalspecification.html#size `End(i) - Start(i) + point-size` 
But pointsize also needs to be 1n then

File: src/datatypes/interval.ts [hunk 16]
```diff
@@ -556,8 +570,8 @@ export class Interval {
       throw new Error('Point type of intervals cannot be determined.');
     }
 
-    if (typeof pointSize === 'number') {
-      pointSize = new Quantity(pointSize, '1');
+    if (typeof pointSize === 'number' || typeof pointSize === 'bigint') {
+      pointSize = new Quantity(Number(pointSize), '1');
```
Comment (lmd59) [inline src/datatypes/interval.ts]: I think ideally we'd want the pointsize for bigint to be 1n. The `value` and `unit` for a Quantity both seem to be of type `any`. Is there a reason not to make the pointsize into a bigint value?

File: src/datatypes/interval.ts
Comment (lmd59) [inline src/datatypes/interval.ts]:
```suggestion
        pointSize = this.high - predecessor(this.high);
```
can overflow as is

File: src/elm/interval.ts [hunk 17]
```diff
@@ -747,7 +783,16 @@ function collapseIntervals(intervals: any, perWidth: any) {
           a = b;
         }
       } else {
-        if (b.low - a.high <= perWidth.value) {
+        const distance = b.low - a.high;
+        const comparablePerWidth =
+          typeof distance === 'bigint' && Number.isInteger(perWidth.value)
```
Comment (lmd59) [inline src/elm/interval.ts:800]: A little confused about this section. What if perWidth is a decimal? Could casting Number(b.low) below produce a precision issue?

File: test/elm/interval/data.cql [hunk 18]
```diff
@@ -90,6 +106,18 @@ define UnknownOpenEndContainsInt: Interval[0, null) contains 0
 define UnknownClosedEndContainsInt: Interval[0, null] contains 0
 define UnknownEndMayContainInt: Interval[0, null) contains 123456789
 define UnknownEndNotContainsInt: Interval[0, null) contains -1
+define NegInfBegContainsLong: Interval[null, 5L] contains -7L
+define NegInfBegNotContainsLong: Interval[null, 5L] contains 7L
+define UnknownOpenBegContainsLong: Interval(null, 5L] contains 5L
+define UnknownClosedBegContainsLong: Interval[null, 5L] contains 5L
```
Comment (lmd59) [inline test/elm/interval/data.cql]: Isn't this neg infinity closed beginning (rather than unknown)?

Comment (lmd59) [reply test/elm/interval/data.cql]: Same for all other define statements in this file starting with `UnknownClosed`

File: test/elm/interval/data.cql [hunk 19]
```diff
@@ -833,6 +1005,7 @@ define DateTimeIntervalEndsStartsFalse: Interval[DateTime(2012, 1, 5), DateTime(
 
 // @Test: IntegerIntervalUnion
```
Comment (lmd59) [inline test/elm/interval/data.cql:1004]: Superoptional, but since "Integer" test sets (i.e. `IntegerIntervalUnion`) now include Long tests, is it worth a slight rename (or separating out Long tests) for clarity? Longs are Integers in reality, so also feel free to ignore this comment

**Review by lmd59 — COMMENTED** [review]
Excellent updates, and addressing a lot of new Long functionality and improved type handling!
I added some comments, but most of them are small!

File: src/datatypes/interval.ts (hunk 13 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: The normal `number`-based representation on line 513 uses `Math.abs(closed.high - closed.low)`. Since `Math.abs` doesn't work with `bigint`, I was using the ternary to get the absolute value instead (and more accurately mirror the number-based representation).

But you're right. Neither is really necessary, because the [authoring guide](https://cql.hl7.org/02-authorsguide.html#interval-values) says:

> Note that the ending boundary must be greater than or equal to the starting boundary to construct a valid interval.

So `closed.high - closed.low` is always a positive value anyway.

File: src/datatypes/interval.ts (hunk 14 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: Same as above, I was using the ternary since `Math.abs` does not work with `bigint` (and we were using `Math.abs` for the `number`-based implementation on line 546). But, like above, this isn't really necessary since the high must always be greater than or equal to the low anyway.

File: src/datatypes/interval.ts (hunk 15 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: Yeah, I think I was doing a shortcut because if we know it is a Long (`bigint`) then we know the point size is `1L` (`1n`). That said, we don't use these types of shortcuts for the other cases, and we've already calculated `pointSize`, so yeah, it probably makes sense to use `pointSize.value` to remain consistent with the others.

File: src/datatypes/interval.ts
Comment (cmoesel) [reply src/datatypes/interval.ts]: Woah. This bug is ancient. Git blame shows it was carried over in the conversion from `.coffee` to `.js`, which was 6 years ago, but who knows how long it was in the `.coffee` before that). Excellent find!

File: src/datatypes/interval.ts (hunk 16 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: You know, I feel like I had a good reason for this -- like the spec didn't define a `Long` return type for size. But I'm looking at it again and most of the spec stuff around this uses `<T>`, so `1n` seems appropriate.

Or maybe this was just a straight-up mistake. Anyway, I agree with you.

File: src/elm/interval.ts (hunk 17 above)
Comment (cmoesel) [reply src/elm/interval.ts:800]: Since the `perWidth` argument is a `Quantity`, then the `perWidth.value` _is_ a Decimal (in terms of CQL). But JS doesn't allow us to compare `number` w/ `bigint`, so we _need_ to convert the `distance` and `perWidth` to like types.

When `perWidth` is an integer (no decimal component), then it's safest to convert the `perWidth` to a `bigint` and compare `bigint` to `bigint` -- because integer always converts cleanly to `bigint`.

But if `perWidth` has decimal components, then it _can't_ be converted to `bigint`. The only choice we have is to convert `bigint` to a `number`. This _could_ introduce precision issues for Longs that are greater than the max safe integer representation in JS, but there's very little we could do. So we're doing the best we can. I guess the alternative would be to throw an error if we detect that we'll lose precision, but the CQL spec doesn't indicate that `collapse` can throw, so this very well might take implementers by surprise.

That said, it might be safer to do `Number(distance)` on line 793 rather than `Number(b.low) - Number(a.high)` since the difference is likely to be a smaller number than `b.low` or `b.high` -- and therefore less likely to run into precision issues. So, yeah, we probably should do that at least.

File: test/elm/aggregate/data.cql (hunk 8 above)
Comment (cmoesel) [reply test/elm/aggregate/data.cql]: Yeah, it's probably better to have at least two arguments. Good catch.

File: test/elm/interval/data.cql (hunk 18 above)
Comment (cmoesel) [reply test/elm/interval/data.cql]: Yeah. Probably a bad copy/paste exacerbated by a series of copy/pastes after.

I think that line 111 should be `UnknownBegContainsLong` and line 112 can be removed completely since we already have a `NegInfContainsLong` case. And similar changes should be made in all the other places with was copy/pasted.

File: test/elm/interval/data.cql (hunk 19 above)
Comment (cmoesel) [reply test/elm/interval/data.cql:1004]: We should probably separate the Long tests out.

**Comment by cmoesel** [comment]
Fantastic feedback again, @lmd59. You're making me feel like I'm getting sloppy. Thanks for the thorough review and great finds. I'll fix these soon (but probably not today).

File: src/datatypes/interval.ts (hunk 15 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: Now I see why I did it this way. `pointSize` is a `Quantity` (which is why we call `.value` on it). `Quantity.value` is a `number` (not a `bigint`) -- so we actually have to use `BigInt(pointSize.value)` if we want to use the existing `pointSize`. I went ahead and did that for consistency.

File: src/datatypes/interval.ts (hunk 16 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: This `getPointSize()` function normalizes everything to a Quantity, probably to make it easy to deal with it in a unifying way (and because before we supported Long, everything could be represented as a `number` or `Quantity` (whose `value` is a `number`).

I'll look at this. If we change it, then the function signature for `getPointSize()` will have to change to return `Quantity | Number | BigInt`. I'll have to see how much havoc that wreaks.

File: test/elm/interval/data.cql (hunk 19 above)
Comment (cmoesel) [reply test/elm/interval/data.cql:1004]: I actually addressed this as part of a rebase that required a bunch of modifications in this file anyway. So once I push the rebased branch, this comment will probably disappear -- but... I did it!

File: test/elm/aggregate/data.cql (hunk 8 above)
Comment (cmoesel) [reply test/elm/aggregate/data.cql]: Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: test/elm/aggregate/data.cql (hunk 9 above)
Comment (cmoesel) [reply test/elm/aggregate/data.cql]: Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: test/elm/aggregate/data.cql (hunk 10 above)
Comment (cmoesel) [reply test/elm/aggregate/data.cql]: Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: test/elm/aggregate/data.cql (hunk 11 above)
Comment (cmoesel) [reply test/elm/aggregate/data.cql]: Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: src/datatypes/interval.ts (hunk 13 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: Fixed in Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18) -- and I went ahead and removed the unnecessary `Math.abs` calls too.

File: src/datatypes/interval.ts (hunk 14 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18) -- and I went ahead and removed the unnecessary `Math.abs` calls too.

File: src/datatypes/interval.ts (hunk 15 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: Actually, after refactoring `getPointSize()`, I was able to make this just use `pointSize` and changed the other number ones to just use `pointSize` too (instead of `pointSize.value`). Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: src/datatypes/interval.ts (hunk 16 above)
Comment (cmoesel) [reply src/datatypes/interval.ts]: Refactored `getPointSize()` to return a `Quantity`, `Number`, or `BigInt` in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: src/datatypes/interval.ts
Comment (cmoesel) [reply src/datatypes/interval.ts]: Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: src/elm/interval.ts (hunk 17 above)
Comment (cmoesel) [reply src/elm/interval.ts:800]: Updated to use `Number(distance)` in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18), which still isn't perfect, but I think is the best we can do.

File: test/elm/interval/data.cql (hunk 18 above)
Comment (cmoesel) [reply test/elm/interval/data.cql]: Fixed in [b36e5ff](https://github.com/cqframework/cql-execution/pull/363/commits/b36e5ff68d680085af9f97c5100efb9286924d18).

File: test/elm/interval/data.cql (hunk 19 above)
Comment (cmoesel) [reply test/elm/interval/data.cql:1004]: Fixed in 5033228 during rebase operation.

**Comment by cmoesel** [comment]
@lmd59 - I have addressed all of your comments (I think) plus a few more things that I discovered along the way. This is ready for re-review.

@dehall - I welcome your review as well.

File: test/elm/arithmetic/data.cql [hunk 20]
```diff
@@ -50,23 +65,34 @@ define DivideNumberByUncertainty: 36 / UncertaintySixToEighteen
...
 define Parenthetical: (1 + 5) * (10 - 15) / 3
 
 // @Test: Power
 define Pow: 3 ^ 4
+define NegPow: 10 ^ -1
+define ThreeExpFourLong: 3L ^ 4L
+define ThreeExpFourMixed: 3 ^ 4L
+define ThreeExpFourReverseMixed: 3L ^ 4
+define TenLongExpNegativeOneLong: 10L ^ -1L
+define TwoLongExpMaxLong: 2L ^ maximum Long
```
Comment (lmd59) [inline test/elm/arithmetic/data.cql:83]: When looking at these new exponent case, I noticed that we don't have a test along the lines of
`define TwoLongExpMinLong: 2L ^ minimum Long`
I would expect this to return null because it can't safely be converted. Is this a case of "the best we can do but not necessarily aligned with spec"? Just wondering if we should add a test for this or otherwise catalog this case?

**Review by lmd59 — COMMENTED** [review]
Just one tiny question - looking great to me! Excited that this will be supported soon!

File: test/elm/arithmetic/data.cql (hunk 20 above)
Comment (cmoesel) [reply test/elm/arithmetic/data.cql:83]: Thanks! `2L ^ minimum Long` does indeed return null. I'm not sure what would be best here: 
* Add a test that asserts the answer is `null` since we can't safely execute that case (in which case the test is validating that we return `null` instead of a completely _wrong_ answer).
* Add a test that asserts 0.00000000 (which I think would be the answer if we did support it due to capping precision at 8 decimals) and then skip it.

What do you think?

Or I can introduce [BigNumber.js](https://github.com/mikemcl/bignumber.js/) (or it's smaller sibling `Big.js`) to handle just this one edge case properly. We might end up using `BigNumber.js` to improve our support for decimals anyway (cc: @dehall).

Comment (lmd59) [reply test/elm/arithmetic/data.cql:83]: I think I would lean toward the 0.00000000 option. It helps acknowledge limitations by skipping a test. If we decide to pull in BigNumber or similar later, we can potentially re-evaluate handling this edge case. But I don't think I'd want to pull in a new dependency for that one unusual edge case at this point.

Comment (cmoesel) [reply test/elm/arithmetic/data.cql:83]: Skipped test added in 1f0cb98.

File: test/elm/arithmetic/arithmetic-test.ts [hunk 21]
```diff
@@ -226,6 +317,35 @@ describe('Power', () => {
...
+
+  it('should be able to calculate the negative power of a long', async function () {
+    (await this.tenLongExpNegativeOneLong.exec(this.ctx)).should.equal(0.1);
+  });
+
+  it('should return null when a long power exponent is too large (beyond max Long value)', async function () {
+    should(await this.twoLongExpMaxLong.exec(this.ctx)).be.null();
+  });
+
+  // TOODO: Unskip this test when we properly handle negative Long exponents that can't be safely converted to Number
```
Comment (lmd59) [inline test/elm/arithmetic/arithmetic-test.ts]:
```suggestion
  // TODO: Unskip this test when we properly handle negative Long exponents that can't be safely converted to Number
```

**Review by lmd59 — APPROVED** [review]

File: test/elm/arithmetic/arithmetic-test.ts (hunk 21 above)
Comment (cmoesel) [reply test/elm/arithmetic/arithmetic-test.ts]: _(sigh)_

File: src/util/math.ts [hunk 22]
```diff
@@ -147,24 +189,25 @@ export function successor(val: any): any {
     // For uncertainties, if the high is the max val, don't increment it
     const high = (() => {
       try {
-        return successor(val.high);
+        return successor(val.high, type);
       } catch {
         return val.high;
       }
     })();
     return new Uncertainty(successor(val.low), high);
```
Comment (dehall) [inline src/util/math.ts]: Should `type` be passed in here as well? Same below for `predecessor`
```suggestion
    return new Uncertainty(successor(val.low, type), high);
```

On a related note, I don't see any tests that seem like they would hit this branch. I wasn't immediately sure what the best place for one was was but this quickly hacked together test (ostensibly `test/util/math-test.ts`) shows the issue:


```ts
import should from 'should';
import { Uncertainty } from '../../src/datatypes/uncertainty';
import { successor, predecessor } from '../../src/util/math';
import { ELM_DECIMAL_TYPE, ELM_INTEGER_TYPE } from '../../src/util/elmTypes';

describe('successor', () => {
  it('should preserve integers in an Uncertainty', () => {
    const result = successor(new Uncertainty(1.0, 2.0), ELM_INTEGER_TYPE);
    result.low.should.equal(2);
    result.high.should.equal(3);
  });
  it('should preserve decimals in an Uncertainty', () => {
    const result = successor(new Uncertainty(1.0, 2.0), ELM_DECIMAL_TYPE);
    result.low.should.equal(1.00000001);
    result.high.should.equal(2.00000001);
  });
});

describe('predecessor', () => {
  it('should preserve integers in an Uncertainty', () => {
    const result = predecessor(new Uncertainty(1.0, 2.0), ELM_INTEGER_TYPE);
    result.low.should.equal(0);
    result.high.should.equal(1);
  });
  it('should preserve decimals in an Uncertainty', () => {
    const result = predecessor(new Uncertainty(1.0, 2.0), ELM_DECIMAL_TYPE);
    result.low.should.equal(0.99999999);
    result.high.should.equal(1.99999999);
  });
});
```

File: src/util/math.ts [hunk 23]
```diff
@@ -199,15 +248,15 @@ export function predecessor(val: any): any {
     // For uncertainties, if the low is the min val, don't decrement it
     const low = ((): any => {
       try {
-        return predecessor(val.low);
+        return predecessor(val.low, type);
       } catch {
         return val.low;
       }
     })();
     return new Uncertainty(low, predecessor(val.high));
```
Comment (dehall) [inline src/util/math.ts]: Same as above - should `type` be passed in here as well?
```suggestion
    return new Uncertainty(low, predecessor(val.high, type));
```

File: src/elm/interval.ts [hunk 24]
```diff
@@ -588,10 +597,42 @@ export class Expand extends Expression {
...
+      if (perBigInt > longHigh - longLow + 1n) {
+        return [];
+      }
+
+      let current_low = longLow;
+      let current_high = current_low + perBigInt - 1n;
+      const results = [];
+      while (current_high <= longHigh) {
+        results.push(new dtivl.Interval(current_low, current_high, true, true));
+        current_low += perBigInt;
```
Comment (dehall) [inline src/elm/interval.ts:632]: I realize it's invalid for the per value on an interval to be 0, but it does appear to be technically possible to reach here with `perBigInt = 0`. 

eg:
```
define ZeroPer: expand { Interval[2 'g', 4 'g'] } per 0 'g'
```

This then becomes an infinite loop (eventually crashing, out of memory, probably because of `results` growing too large). I can't find anything obvious in the spec on what to do in this situation. https://cql.hl7.org/09-b-cqlreference.html#expand does mention `null` as a possibility but not 0. Since there are already a few early exits to this function that `return []`, if there's no formal guidance on what to do I'd propose just adding a 0 check in there as well, but I'm open to anything.

File: src/elm/type.ts [hunk 25]
```diff
@@ -173,6 +240,8 @@ export class ToQuantity extends Expression {
       return null;
     } else if (typeof val === 'number') {
       return new Quantity(val, '1');
+    } else if (typeof val === 'bigint') {
+      return new Quantity(Number(val), '1');
```
Comment (dehall) [inline src/elm/type.ts:248]: Super minor: you could have a comment here that converting a BigInt to a Number does theoretically have some loss of precision but because the range of safe integers is so big*, it's probably not a big deal in the context of CQL Quantity

\* = `-(2^53 - 1) to 2^53 - 1, (±9,007,199,254,740,991)`

(aka plus or minus 9 quadrillion, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger#description )

**Review by dehall — CHANGES_REQUESTED** [review]
This looks good. I came at this primarily from the perspective of how one might build off this for a future "BigDecimal"-based type and I feel like this is a good model to follow. One or maybe two spots I might have refactored some duplicate logic into a helper function but mixing JS primitives with classes means that's not really possible very often.
 
In the interest of understanding what our modern AI tools are capable of I did ask Codex to see what it thought, and it highlighted a few items which I confirmed. I think the 2 Uncertainty code suggestions I made are necessary but I defer to you on what you want to do for everything else

File: src/util/math.ts (hunk 22 above)
Comment (cmoesel) [reply src/util/math.ts]: Good catch. I thought I did a global find/replace for `successor`/`predecessor`, but apparently I missed some. I also thought I did a full Codex review of my own code, but maybe I made the `successor`/`predecessor` change after that (or maybe Codex just gave you different findings than it gave me, which I've definitely seen before).

File: src/elm/type.ts (hunk 25 above)
Comment (cmoesel) [reply src/elm/type.ts:248]: If I'm making changes in the code already, then I might as well. This will likely become a moot point once we introduce a class to handle Decimals better.

File: src/elm/interval.ts (hunk 24 above)
Comment (cmoesel) [reply src/elm/interval.ts:632]: Good catch. I think `per 0 'g'` is basically just a variation of the divide-by-zero problem. As such, I think it makes sense to return `null` since that's what [CQL Divide](https://cql.hl7.org/09-b-cqlreference.html#divide) specifies for divide-by-zero.

File: src/util/math.ts (hunk 22 above)
Comment (cmoesel) [reply src/util/math.ts]: Fixed in [ede09db](https://github.com/cqframework/cql-execution/pull/363/commits/ede09db3cf70a36d7e80ff76e5e4a57a660e2aad) (with a few other bonus successor/predecessor fixes to add the type).

File: src/util/math.ts (hunk 23 above)
Comment (cmoesel) [reply src/util/math.ts]: Fixed in [ede09db](https://github.com/cqframework/cql-execution/pull/363/commits/ede09db3cf70a36d7e80ff76e5e4a57a660e2aad) (with a few other bonus successor/predecessor fixes to add the type).

File: src/elm/interval.ts (hunk 24 above)
Comment (cmoesel) [reply src/elm/interval.ts:632]: Fixed in [5942e0b](https://github.com/cqframework/cql-execution/pull/363/commits/5942e0bbc953b99449fc451527406427fe98b740).

File: src/elm/type.ts (hunk 25 above)
Comment (cmoesel) [reply src/elm/type.ts:248]: Added in [c181ffa](https://github.com/cqframework/cql-execution/pull/363/commits/c181ffade8ca7fae05228f904d9f980aed0ed36e).

**Comment by cmoesel** [comment]
Thanks for the review @dehall. I've addressed your comments.

I was tempted to add a whole new suite of decimal tests in the interval tests (like the `IntegerInterval` and `LongInterval` suites), but I decided against it since it is slightly outside the scope of this PR and since we'll probably be updating Decimal support soon anyway.

**Review by dehall — APPROVED** [review]
